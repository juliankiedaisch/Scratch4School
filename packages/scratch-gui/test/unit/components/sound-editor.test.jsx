import React from 'react';
import {renderWithIntl} from '../../helpers/intl-helpers.jsx';
import SoundEditor from '../../../src/components/sound-editor/sound-editor';
import {fireEvent, waitFor} from '@testing-library/react';

describe('Sound Editor Component', () => {
    let props;
    beforeEach(() => {
        props = {
            canUndo: true,
            canRedo: true,
            chunkLevels: [1, 2, 3],
            name: 'sound name',
            playhead: 0.5,
            trimStart: 0.2,
            trimEnd: 0.8,
            onChangeName: jest.fn(),
            onDelete: jest.fn(),
            onPlay: jest.fn(),
            onRedo: jest.fn(),
            onReverse: jest.fn(),
            onSofter: jest.fn(),
            onLouder: jest.fn(),
            onRobot: jest.fn(),
            onEcho: jest.fn(),
            onFaster: jest.fn(),
            onSlower: jest.fn(),
            onSetTrimEnd: jest.fn(),
            onSetTrimStart: jest.fn(),
            onStop: jest.fn(),
            onUndo: jest.fn()
        };
    });

    test('matches snapshot', () => {
        const {container} = renderWithIntl(<SoundEditor {...props} />);

        expect(container.firstChild).toMatchSnapshot();
    });

    test('delete button appears when selection is not null', () => {
        const {container} = renderWithIntl(
            <SoundEditor
                {...props}
                trimEnd={0.75}
                trimStart={0.25}
            />
        );
        const deleteButton = [...container.querySelectorAll('div[role="button"]')]
            .find(el => el.textContent.trim() === 'Delete');

        fireEvent.click(deleteButton);
        expect(props.onDelete).toHaveBeenCalled();
    });

    test('play button appears when playhead is null', () => {
        const {container} = renderWithIntl(
            <SoundEditor
                {...props}
                playhead={null}
            />
        );
        const playButton = container.querySelector('button[title="Play"]');
        fireEvent.click(playButton);
        expect(props.onPlay).toHaveBeenCalled();
    });

    test('stop button appears when playhead is not null', () => {
        const {container} = renderWithIntl(
            <SoundEditor
                {...props}
                playhead={0.5}
            />
        );
        const stopButton = container.querySelector('button[title="Stop"]');
        fireEvent.click(stopButton);
        expect(props.onStop).toHaveBeenCalled();
    });

    test('submitting name calls the callback', async () => {
        if (typeof MutationObserver === 'undefined') {
            global.MutationObserver = class {
                observe () { }
                disconnect () { }
            };
        }
        const onChangeName = jest.fn();
        const {container} = renderWithIntl(<SoundEditor
            {...props}
            onChangeName={onChangeName}
        />);

        const input = container.querySelector('input');

        fireEvent.change(input, {target: {value: 'hello'}});
        fireEvent.keyPress(input, {key: 'Enter', code: 'Enter', charCode: 13});
        await waitFor(() => expect(onChangeName).toHaveBeenCalled());
    });

    describe('effect buttons call the correct callbacks', () => {
        let getButtonByText;

        beforeEach(() => {
            const {container} = renderWithIntl(
                <SoundEditor {...props} />
            );

            const buttons = [...container.querySelectorAll('div[role="button"]')];
            getButtonByText = text => buttons.find(div => div.textContent.trim() === text);
        });

        test('clicking reverse button calls correct callback', () => {
            const reverseButton = getButtonByText('Reverse');
            fireEvent.click(reverseButton);
            expect(props.onReverse).toHaveBeenCalled();
        });

        test('clicking robot button calls correct callback', () => {
            const robotButton = getButtonByText('Robot');
            fireEvent.click(robotButton);
            expect(props.onRobot).toHaveBeenCalled();
        });

        test('clicking faster button calls correct callback', () => {
            const fasterButton = getButtonByText('Faster');
            fireEvent.click(fasterButton);
            expect(props.onFaster).toHaveBeenCalled();
        });

        test('clicking slower button calls correct callback', () => {
            const slowerButton = getButtonByText('Slower');
            fireEvent.click(slowerButton);
            expect(props.onSlower).toHaveBeenCalled();
        });

        test('clicking louder button calls correct callback', () => {
            const louderButton = getButtonByText('Louder');
            fireEvent.click(louderButton);
            expect(props.onLouder).toHaveBeenCalled();
        });

        test('clicking softer button calls correct callback', () => {
            const softerButton = getButtonByText('Softer');
            fireEvent.click(softerButton);
            expect(props.onSofter).toHaveBeenCalled();
        });
    });

    describe('disbaling undo/redo button', () => {
        test('undo button can be disabled when canUndo equals false', () => {
            const {container} = renderWithIntl(
                <SoundEditor
                    {...props}
                    canUndo={false}
                />
            );

            const undoButtonEnabled = container.querySelector('button[title="Undo"]');
            expect(undoButtonEnabled.disabled).toBe(true);
        });

        test('redo button can be disabled when canRedo equals false', () => {
            const {container} = renderWithIntl(
                <SoundEditor
                    {...props}
                    canRedo={false}
                />
            );

            const redoButtonEnabled = container.querySelector('button[title="Redo"]');
            expect(redoButtonEnabled.disabled).toBe(true);
        });
    });

    describe('undo/redo buttons call the correct callbacks', () => {
        test('when undo button is clicked it calls the correct callback', () => {
            const {container} = renderWithIntl(
                <SoundEditor {...props} />
            );

            const undoButton = container.querySelector('button[title="Undo"]');
            fireEvent.click(undoButton);
            expect(props.onUndo).toHaveBeenCalled();
        });

        test('when redo button is clicked it calls the correct callback', () => {
            const {container} = renderWithIntl(
                <SoundEditor {...props} />
            );

            const redoButton = container.querySelector('button[title="Redo"]');
            fireEvent.click(redoButton);
            expect(props.onRedo).toHaveBeenCalled();
        });
    });
});
