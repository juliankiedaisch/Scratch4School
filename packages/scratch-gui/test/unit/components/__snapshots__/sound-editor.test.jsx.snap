// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sound Editor Component matches snapshot 1`] = `
<div>
  <div>
    <div>
      <label>
        <span>
          Sound
        </span>
        <input
          class=""
          tabindex="1"
          type="text"
          value="sound name"
        />
      </label>
      <div>
        <button
          title="Undo"
        >
          <img
            draggable="false"
            src="test-file-stub"
          />
        </button>
        <button
          title="Redo"
        >
          <img
            draggable="false"
            src="test-file-stub"
          />
        </button>
      </div>
    </div>
    <div>
      <div
        class=""
        role="button"
      >
        <img
          draggable="false"
          src="test-file-stub"
        />
        <div>
          Copy
        </div>
      </div>
      <div
        class=""
        role="button"
      >
        <img
          draggable="false"
          src="test-file-stub"
        />
        <div>
          Paste
        </div>
      </div>
      <div
        class=""
        role="button"
      >
        <img
          draggable="false"
          src="test-file-stub"
        />
        <div>
          Copy to New
        </div>
      </div>
    </div>
    <div
      class=""
      role="button"
    >
      <img
        draggable="false"
        src="test-file-stub"
      />
      <div>
        Delete
      </div>
    </div>
  </div>
  <div>
    <div>
      <svg
        viewBox="0 0 600 160"
      >
        <g
          transform="scale(1, -1) translate(0, -80)"
        >
          <path
            d="M0 0Q0 80 150 120 Q300 160 450 200 Q600 240 600 0 Q600 -240 450 -200 Q300 -160 150 -120 Q0 -80 0 0Z"
            stroke-linejoin="round"
            stroke-width="1"
          />
        </g>
      </svg>
      <div
        class=""
      >
        <div
          class=""
          style="width: 60.00000000000001%; left: 20%;"
        >
          <div
            class=""
          />
          <div
            class=""
          >
            <div
              class=""
            >
              <img
                draggable="false"
                src="test-file-stub"
              />
            </div>
            <div
              class=""
            >
              <img
                draggable="false"
                src="test-file-stub"
              />
            </div>
          </div>
          <div
            class=""
          >
            <div
              class=""
            >
              <img
                draggable="false"
                src="test-file-stub"
              />
            </div>
            <div
              class=""
            >
              <img
                draggable="false"
                src="test-file-stub"
              />
            </div>
          </div>
        </div>
        <div>
          <div
            class=""
            style="transform: translateX(50%);"
          />
        </div>
      </div>
    </div>
  </div>
  <div
    class=""
  >
    <div>
      <button
        class=""
        title="Stop"
      >
        <img
          draggable="false"
          src="test-file-stub"
        />
      </button>
    </div>
    <div
      class=""
      role="button"
    >
      <img
        draggable="false"
        src="test-file-stub"
      />
      <div>
        Faster
      </div>
    </div>
    <div
      class=""
      role="button"
    >
      <img
        draggable="false"
        src="test-file-stub"
      />
      <div>
        Slower
      </div>
    </div>
    <div
      class=""
      role="button"
    >
      <img
        draggable="false"
        src="test-file-stub"
      />
      <div>
        Louder
      </div>
    </div>
    <div
      class=""
      role="button"
    >
      <img
        draggable="false"
        src="test-file-stub"
      />
      <div>
        Softer
      </div>
    </div>
    <div
      class=""
      role="button"
    >
      <img
        draggable="false"
        src="test-file-stub"
      />
      <div>
        Mute
      </div>
    </div>
    <div
      class=""
      role="button"
    >
      <img
        draggable="false"
        src="test-file-stub"
      />
      <div>
        Fade in
      </div>
    </div>
    <div
      class=""
      role="button"
    >
      <img
        draggable="false"
        src="test-file-stub"
      />
      <div>
        Fade out
      </div>
    </div>
    <div
      class=""
      role="button"
    >
      <img
        draggable="false"
        src="test-file-stub"
      />
      <div>
        Reverse
      </div>
    </div>
    <div
      class=""
      role="button"
    >
      <img
        draggable="false"
        src="test-file-stub"
      />
      <div>
        Robot
      </div>
    </div>
  </div>
</div>
`;
